### PR TITLE
Fixes unscoping the where clause when it has an array `default_scope`…

### LINKED ISSUE #30967: ActiveRecord does not unscope where conditions when right-hand side is a group
### Steps to reproduce
https://gist.github.com/jemminger/b14dc10bc1501c9efce081729182de33

### Expected behavior
Calling `unscope` for `:where` on a column with a group RH side should remove that column's conditions

### Actual behavior
The column's conditions remain

### System configuration
**Rails version**: 5.1.3

**Ruby version**: 2.3.4

### PR DESCRIPTION
**Pull request by jemminger**
### Summary

Fixes https://github.com/rails/rails/issues/30967

### CODE REVIEW

File: activerecord/test/models/post.rb [hunk 1]
```diff
@@ -233,6 +233,12 @@ class PostWithDefaultScope < ActiveRecord::Base
   default_scope { order(:title) }
 end
 
+class PostWithArrayDefaultScope < ActiveRecord::Base
```
[the remark this answers is not in the document]
Comment (jemminger) [reply activerecord/test/models/post.rb]: In my testing, the bug only arises when the model has the array where condition in a default_scope.

File: activerecord/test/cases/scoping/default_scoping_test.rb [hunk 2]
```diff
@@ -422,6 +422,11 @@ def test_unscoped_with_joins_should_not_have_default_scope
                  Comment.joins(:post).to_a
   end
 
+  def test_unscoping_an_array_default_column
```
[the remark this answers is not in the document]
Comment (jemminger) [reply activerecord/test/cases/scoping/default_scoping_test.rb]: In my testing, the bug only arises when the model has the array where condition in a default_scope.

File: activerecord/lib/active_record/relation/where_clause.rb [hunk 3]
```diff
@@ -139,11 +141,24 @@ def invert_predicate(node)
         end
 
         def except_predicates(columns)
+          relevant_nodes = [Arel::Nodes::Between, Arel::Nodes::In, Arel::Nodes::NotIn, Arel::Nodes::Equality, Arel::Nodes::NotEqual, Arel::Nodes::LessThan, Arel::Nodes::LessThanOrEqual, Arel::Nodes::GreaterThan, Arel::Nodes::GreaterThanOrEqual]
```
[the remark this answers is not in the document]
Comment (jemminger) [reply activerecord/lib/active_record/relation/where_clause.rb]: Yes, I misunderstood and made the node tester a constant.  Will fix

File: activerecord/lib/active_record/relation/where_clause.rb [hunk 4]
```diff
@@ -3,6 +3,12 @@
 module ActiveRecord
   class Relation
     class WhereClause # :nodoc:
+
+      NODE_TESTER = -> (node, columns) {
```
[the remark this answers is not in the document]
Comment (jemminger) [reply activerecord/lib/active_record/relation/where_clause.rb]: Will fix

File: activerecord/lib/active_record/relation/where_clause.rb [hunk 5]
```diff
@@ -139,11 +141,24 @@ def invert_predicate(node)
...
           predicates.reject do |node|
             case node
-            when Arel::Nodes::Between, Arel::Nodes::In, Arel::Nodes::NotIn, Arel::Nodes::Equality, Arel::Nodes::NotEqual, Arel::Nodes::LessThan, Arel::Nodes::LessThanOrEqual, Arel::Nodes::GreaterThan, Arel::Nodes::GreaterThanOrEqual
-              subrelation = (node.left.kind_of?(Arel::Attributes::Attribute) ? node.left : node.right)
-              columns.include?(subrelation.name.to_s)
+            when *relevant_nodes
+              NODE_TESTER.call(node, columns)
+
+            when Arel::Nodes::Grouping
+              node.each do |n|
```
[the remark this answers is not in the document]
Comment (jemminger) [reply activerecord/lib/active_record/relation/where_clause.rb]: Will do

**Comment by jemminger** [comment]
@gmcgibbon Can you take a quick look at this?  I don't think my changes are causing the CI failures.

**Comment by adbatista** [comment]
@gmcgibbon I'm having the same problem, is there a chance of this being released any time soon?

**Comment by stanleeDK** [comment]
Hi Alessandro,

Would you mind removing Stan Lee from this repo. He no longer works at
BuildZoom, but his github is associated with this email and we haven't been
able to get him unsubscribed so we are getting frequent mails to our
catchall.

Thanks!

File: activerecord/lib/active_record/relation/where_clause.rb [hunk 6]
```diff
@@ -141,13 +139,20 @@ def invert_predicate(node)
         def except_predicates(columns)
           predicates.reject do |node|
             case node
-            when Arel::Nodes::Between, Arel::Nodes::In, Arel::Nodes::NotIn, Arel::Nodes::Equality, Arel::Nodes::NotEqual, Arel::Nodes::LessThan, Arel::Nodes::LessThanOrEqual, Arel::Nodes::GreaterThan, Arel::Nodes::GreaterThanOrEqual
-              subrelation = (node.left.kind_of?(Arel::Attributes::Attribute) ? node.left : node.right)
-              columns.include?(subrelation.name.to_s)
+            when *EXCEPT_PREDICATE_NODES
```
Comment (gmcgibbon) [inline activerecord/lib/active_record/relation/where_clause.rb]: This works, but I think it obfuscates the real problem. `Arel.fetch_attribute` doesn't parse attributes from node groups correctly. You can do something like this to fix that more directly with recursion:
```diff
--- a/activerecord/lib/arel.rb
+++ b/activerecord/lib/arel.rb
@@ -46,10 +46,18 @@ def self.arel_node?(value) # :nodoc:
     value.is_a?(Arel::Node) || value.is_a?(Arel::Attribute) || value.is_a?(Arel::Nodes::SqlLiteral)
   end

-  def self.fetch_attribute(value) # :nodoc:
+  def self.fetch_attribute(value, &block) # :nodoc:
     case value
-    when Arel::Nodes::Between, Arel::Nodes::In, Arel::Nodes::NotIn, Arel::Nodes::Equality, Arel::Nodes::NotEqual, Arel::Nodes
::LessThan, Arel::Nodes::LessThanOrEqual, Arel::Nodes::GreaterThan, Arel::Nodes::GreaterThanOrEqual
-      yield value.left.is_a?(Arel::Attributes::Attribute) ? value.left : value.right
+    when Arel::Nodes::Between, Arel::Nodes::In, Arel::Nodes::NotIn, Arel::Nodes::Equality,
+         Arel::Nodes::NotEqual, Arel::Nodes::LessThan, Arel::Nodes::LessThanOrEqual,
+         Arel::Nodes::GreaterThan, Arel::Nodes::GreaterThanOrEqual
+      if value.left.is_a?(Arel::Attributes::Attribute)
+        block.call(value.left)
+      else
+        block.call(value.right)
+      end
+    when Arel::Nodes::Grouping
+      fetch_attribute(value.expr.left, &block) unless value.expr.is_a?(Arel::Nodes::SqlLiteral)
     end
   end
```

File: activerecord/test/cases/scoping/default_scoping_test.rb (hunk 2 above)
Comment (gmcgibbon) [inline activerecord/test/cases/scoping/default_scoping_test.rb]: I think this is better put in `relations_test`, you also don't need a subclass or scope:

```ruby
def test_unscope_grouped_where
  posts = Post.where(
    title: ["Welcome to the weblog", "So I was thinking", nil]
  )

  assert_equal 2, posts.count
  assert_equal Post.count, posts.unscope(where: :title).count
end
```

**Review by gmcgibbon — CHANGES_REQUESTED** [review]

**Pull request closed without merging**